onmt/utils/loss.py

```diff
@@ -248,10 +250,30 @@ def _make_shard_state(self, batch, output, range_, attns=None):
                 "std_attn": attns.get("std"),
                 "coverage_attn": coverage
             })
+        if self.lambda_align != 0.0:
+            # attn_align should be in (batch_size, pad_tgt_size, pad_src_size)
+            attn_align = attns.get("align", None)
+            # align_idx should be a Tensor in size([N, 3]) of [B, tgt+1, src]
```

Comment:
?

Reply:
batch.align contains alignments in the format of alignment idx rather than alignment matrix. batch.align may contain N pair of src-tgt alignment, each represented as (sentence_number_in_batch, aligned_tgt_id+1, aligned_src_id) . aligned_tgt_id+1 behaves as adding BOS in each sentence, aims to making it coherent with batch.tgt after calling make_batch_align_matrix for batch.align.